File: backend/dotnet/samples/Mateu.Demo/SalesDashboard.cs
using Mateu.Uidl;

namespace Mateu.Demo;

// A declarative dashboard (the C# analogue of the Java Dashboard archetype):
// consecutive MetricCard properties become a Scoreboard KPI band, [Panel] component
// properties become titled tiles on a responsive grid.
[UI("dashboard"), Title("Sales dashboard")]
public class SalesDashboard : Dashboard
{
    public MetricCard Revenue { get; } = new()
    {
        Title = "Revenue", Value = "1.2", Unit = "M€",
        Trend = MetricTrend.Up, TrendLabel = "+12% vs last month", Icon = "💶",
        ActionId = "openRevenue",
    };

    public MetricCard Orders { get; } = new()
    {
        Title = "Orders", Value = "348", Trend = MetricTrend.Neutral, TrendLabel = "flat", Icon = "📦",
    };

    public MetricCard Refunds { get; } = new()
    {
        Title = "Refunds", Value = "7", Trend = MetricTrend.Down, TrendLabel = "-3 this week", Icon = "↩️",
    };

    [Panel(Title = "Delivery plan", Subtitle = "Q3 roadmap", ColSpan = 2)]
    public Gantt Plan { get; } = new()
    {
        Tasks =
        [
            new GanttTask
            {
                Id = "t1", Title = "Discovery", Start = new DateOnly(2026, 7, 1),
                End = new DateOnly(2026, 7, 15), Progress = 100, Color = "#4caf50",
            },
            new GanttTask
            {
                Id = "t2", Title = "Build", Start = new DateOnly(2026, 7, 10),
                End = new DateOnly(2026, 8, 20), Progress = 40,
            },
            new GanttTask
            {
                Id = "t3", Title = "Rollout", Start = new DateOnly(2026, 8, 15),
                End = new DateOnly(2026, 9, 10), Progress = 0,
            },
        ],
    };

    [Panel(Title = "Alerts")]
    public EmptyState Alerts { get; } = new()
    {
        Icon = "🎉", Title = "No alerts", Description = "Everything is running smoothly.",
        ActionId = "refresh", ActionLabel = "Refresh",
    };

    [Panel(Title = "Loading preview")]
    public Skeleton Loading { get; } = new() { Variant = SkeletonVariant.Card, Count = 2 };

    // Drill-in actions referenced by the tiles above.
    public Message OpenRevenue() => new("Revenue drill-in");
    public Message Refresh() => new("Refreshed");
}

// A declarative Redwood-style foldout page: the first component property is the always-visible
// overview; [Panel] properties are lateral fold-out panels.
[UI("booking"), Title("Booking 2026/0042")]
public class BookingFoldout : Foldout
{
    public Text Overview { get; } = new("Booking 2026/0042 — Miguel Pérez — 2 nights, Deluxe room");

    [Panel(Title = "Guests", Icon = "👥")]
    public Text Guests { get; } = new("Miguel Pérez (adult), Ana Pérez (adult)");

    [Panel(Title = "Payments", Icon = "💳", Open = false)]
    public Text Payments { get; } = new("Deposit 120€ received. Balance due at check-in.");

    [Panel(Title = "History", Icon = "🕓", Open = false)]
    public Text History { get; } = new("Created 2026-06-01, modified 2026-06-20.");
}

// A welcome page: hero + call-to-action buttons + highlight tiles.
[UI("welcome"), Title("Welcome")]
public class WelcomeDemo : Welcome
{
    protected override string? HeroTitle => "Welcome to Mateu.NET";
    protected override string? HeroSubtitle => "The same renderers, now speaking C#";

    public Button GetStarted { get; } = new("Get started", "start") { Primary = true };

    [Panel(Title = "Dashboards")]
    public Text Highlight1 { get; } = new("Scoreboards, metric cards, Gantt charts and tiles.");

    [Panel(Title = "Foldouts")]
    public Text Highlight2 { get; } = new("Redwood-style record pages with lateral panels.");

    public Message Start() => new("Let's go!");
}

File: demo/demo-redwood-showcase/src/main/java/io/mateu/mdd/redwoodshowcase/ui/dashboard/SalesDashboard.java
package io.mateu.mdd.redwoodshowcase.ui.dashboard;

import io.mateu.core.infra.declarative.orchestrators.dashboard.Dashboard;
import io.mateu.uidl.annotations.Action;
import io.mateu.uidl.annotations.Panel;
import io.mateu.uidl.annotations.Title;
import io.mateu.uidl.annotations.UI;
import io.mateu.uidl.data.Chart;
import io.mateu.uidl.data.ChartData;
import io.mateu.uidl.data.ChartDataset;
import io.mateu.uidl.data.ChartType;
import io.mateu.uidl.data.Markdown;
import io.mateu.uidl.data.MetricCard;
import io.mateu.uidl.data.MetricTrend;
import java.net.URI;
import java.util.List;

/**
 * Demo of the declarative {@link Dashboard} archetype: consecutive {@link MetricCard} fields become
 * a scoreboard band, and {@code @Panel} component fields become titled tiles on a responsive grid.
 */
@UI("/dashboard-demo")
@Title("Sales dashboard")
public class SalesDashboard extends Dashboard {

  MetricCard revenue =
      MetricCard.builder()
          .title("Revenue")
          .value("1.2")
          .unit("M€")
          .trend(MetricTrend.up)
          .trendLabel("+8% vs last month")
          .icon("vaadin:dollar")
          .build();

  MetricCard orders =
      MetricCard.builder()
          .title("Orders")
          .value("3,421")
          .trend(MetricTrend.up)
          .trendLabel("+112")
          .icon("vaadin:cart")
          .actionId("openOrders")
          .build();

  MetricCard returns =
      MetricCard.builder()
          .title("Returns")
          .value("87")
          .trend(MetricTrend.down)
          .trendLabel("-12% vs last month")
          .icon("vaadin:refresh")
          .build();

  MetricCard nps =
      MetricCard.builder()
          .title("NPS")
          .value("62")
          .trend(MetricTrend.neutral)
          .trendLabel("stable")
          .description("Net promoter score, last 30 days")
          .build();

  @Panel(title = "Monthly sales", subtitle = "Units sold per month", colSpan = 2)
  Chart sales =
      Chart.builder()
          .chartType(ChartType.bar)
          .chartData(
              ChartData.builder()
                  .labels(List.of("Jan", "Feb", "Mar", "Apr", "May", "Jun"))
                  .datasets(
                      List.of(
                          ChartDataset.builder()
                              .label("2026")
                              .data(List.of(120d, 190d, 300d, 250d, 220d, 310d))
                              .build()))
                  .build())
          .build();

  @Panel(title = "Channel mix")
  Chart channels =
      Chart.builder()
          .chartType(ChartType.pie)
          .chartData(
              ChartData.builder()
                  .labels(List.of("Web", "Stores", "Partners"))
                  .datasets(
                      List.of(
                          ChartDataset.builder()
                              .label("Share")
                              .data(List.of(55d, 30d, 15d))
                              .build()))
                  .build())
          .build();

  @Action
  Object openOrders() {
    return URI.create("/products");
  }

  @Panel(title = "Notes")
  Markdown notes =
      new Markdown(
          """
          - Summer campaign starts **July 15th**
          - Stores channel under review
          - Partner onboarding paused
          """,
          null,
          null);
}
